### PR TITLE
Use straight-line propagation for "stiff" tracks  to check GeomDet compatibility

### PR DESCRIPTION
**Pull request by VinInn**
It is well known that about half of the propagation to a tracker sensor fails in particular in
TrajectorySegmentBuilder::redoMeasurements.
This is due to the geometrical topology of Strip sub-layers where only half of the surface is covered by sensors (even less in the endcap)
To mitigate the cost of these failed propagations we first use a straight-line approximation and use that to estimate compatibility with the given GeomDet if the "sagita" is below a given threshold.
The sagita threshold, as well as the minimal tolerance,  are configurable parameters of the MeasurementEstimator.
Parameters has be tuned to minimize efficiency losses... (a better tune is surely possible)

at PU35 for ttbar events we save 35% in GeomDetCompatibilityChecker::isCompatible itself while
TrajectorySegmentBuilder::redoMeasurements becomes a factor 2 faster.

I do expect some iterations to finalize the parameter setting (and fix misspelling)

p.s.
most of the changes are due to the migration of Chi2MeasurementEstimator to fillDescriptions...

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild, please test

**Comment by slava77** [comment]
Hi Vincenzo,

``` I do expect some iterations to finalize the parameter setting (and fix misspelling) ```
Does this mean you will still be updating the PR, or do you need some inputs first (from jenkins or from review)?

**Comment by VinInn** [comment]
@slava, I expect feedback from jenkins and review.
Now I work around HLT config. to be fixed later...

**Comment by slava77** [comment]
@VinInn please refrain from using jenkins for debugging

**Comment by VinInn** [comment]
@cmsbuild, please test

File: RecoTracker/MeasurementDet/plugins/Chi2ChargeMeasurementEstimatorESProducer.cc [hunk 1]
```diff
@@ -102,32 +103,36 @@ namespace {
...
+  boost::shared_ptr<Chi2MeasurementEstimatorBase> m_estimator;
+  const edm::ParameterSet m_pset;
+};
 
-  double maxChi2_;
-  double nSigma_;
-  float minGoodPixelCharge_; 
-  float minGoodStripCharge_;
-  float pTChargeCutThreshold_;
+/*  need support for RefToPSet
```
Comment (slava77) [inline RecoTracker/MeasurementDet/plugins/Chi2ChargeMeasurementEstimatorESProducer.cc]: is this support on the way?
If it's expected some time soon, it makes sense to keep this commented out code

@Dr15Jones

File: RecoTracker/MeasurementDet/plugins/Chi2ChargeMeasurementEstimatorESProducer.cc [hunk 2]
```diff
@@ -135,10 +140,18 @@ Chi2ChargeMeasurementEstimatorESProducer::~Chi2ChargeMeasurementEstimatorESProdu
 boost::shared_ptr<Chi2MeasurementEstimatorBase> 
 Chi2ChargeMeasurementEstimatorESProducer::produce(const TrackingComponentsRecord & iRecord){ 
 
-  _estimator = boost::shared_ptr<Chi2MeasurementEstimatorBase>(
-	new Chi2ChargeMeasurementEstimator(maxChi2_,nSigma_, 
-		minGoodPixelCharge_, minGoodStripCharge_, pTChargeCutThreshold_));
-  return _estimator;
+  auto maxChi2             =  m_pset.getParameter<double>("MaxChi2");
+  auto nSigma              =  m_pset.getParameter<double>("nSigma");
+  auto maxSag  = m_pset.existsAs<double>("MaxSagita") ? m_pset.getParameter<double>("MaxSagita") : -1.;
```
Comment (slava77) [inline RecoTracker/MeasurementDet/plugins/Chi2ChargeMeasurementEstimatorESProducer.cc]: sagitta

File: TrackingTools/DetLayers/src/GeomDetCompatibilityChecker.cc [hunk 3]
```diff
@@ -1,15 +1,111 @@
...
+      if (close) { 
+         stat.nth++;
+         auto pos = plane.toLocal(GlobalPoint(crossing.position(path.second)));
+         // auto toll = LocalError(tolerance2,0,tolerance2);
+         auto tollL2 = std::max(sagita*sagita,minTol2);
+         auto toll = LocalError(tollL2,0,tollL2);
+         isIn = plane.bounds().inside(pos,toll);
+         if (!isIn) { stat.ns2++;
+                      return std::make_pair( false,TrajectoryStateOnSurface());  
+                      // if (!largeErr) return std::make_pair( false,TrajectoryStateOnSurface());
```
Comment (slava77) [inline TrackingTools/DetLayers/src/GeomDetCompatibilityChecker.cc]: is this commented out code needed?

Comment (VinInn) [inline TrackingTools/DetLayers/src/GeomDetCompatibilityChecker.cc]: not anymore..

File: RecoTracker/MeasurementDet/plugins/Chi2ChargeMeasurementEstimatorESProducer.cc (hunk 2 above)
Comment (VinInn) [inline RecoTracker/MeasurementDet/plugins/Chi2ChargeMeasurementEstimatorESProducer.cc]: thanks, will wait for the end of the review

**Comment by slava77** [comment]
I'm done with the code visual inspection and also checked static analysis. No additional comments so far.

There are quite a few changes in comparisons.
Do we have MTV plots?

**Comment by VinInn** [comment]
MTV plots w/r/t pre4  1000 events TTBAR PU35 
http://innocent.home.cern.ch/innocent/RelVal/pu35_804_failProp/

**Comment by slava77** [comment]
- impact on general tracks is pretty small

Most significant differences are in conversions and related:
- reduction in signal-like conversions is visible in many samples, the background end goes up a bit (NB: the "conversions" collection is not really used directly downstream); e.g. gamma35
all_sign659vsorig_singlegamma13pt35wf1319p0c_recoconversions_conversions__reco_obj_mvaout
    - pfPhoton validation confirms loss of efficiency in 2-track conversions for the same sample gamma35
wf1319_conv_eff_2tk_eta
the impact is much smaller with single-leg conversions added
wf1319_conv_eff_eta
It looks like the loss of conversions starts from ckfInOutTracksFromConversions/ckfOutInTracksFromConversions:
in gamma35  (there are more in overflows, but lower radius is at loss)
wf1319_conv_io_r
in ttbar 25202
wf25202_conv_io_r

- conversionStepTracks yield is up by 10% or more on average, more significant at higher pt and in "jetty" samples: e.g. ttbar 25202 (pu35@25ns)
all_sign659vsorig_ttbarpuwf25202p0c_log10recotracks_conversionsteptracks__reco_obj_pt
Note that conversionStepTracks collection is just a part of the inputs to conversion reco, and it is empty in ele/photon gun samples, likely due to vertex selection.

**Comment by slava77** [comment]
timing in ttbar PU35@25ns  (25202 ) goes down primarily due to tracking *Candidates modules
```
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
   -0.834965      -0.17%       217.28 ms/ev ->        89.29 ms/ev convTrackCandidates
   -0.676659      -0.21%       310.64 ms/ev ->       153.58 ms/ev jetCoreRegionalStepTrackCandidates
   -0.244137      -0.07%       259.50 ms/ev ->       203.04 ms/ev conversionTrackCandidates
   -0.221076      -0.09%       345.31 ms/ev ->       276.57 ms/ev tobTecStepTrackCandidates
   -0.140822      -0.11%       657.12 ms/ev ->       570.67 ms/ev lowPtTripletStepTrackCandidates
   -0.136726      -0.09%       549.90 ms/ev ->       479.53 ms/ev pixelLessStepTrackCandidates
   -0.126121      -0.11%       729.66 ms/ev ->       643.10 ms/ev detachedTripletStepTrackCandidates
   -0.078636      -0.08%       780.68 ms/ev ->       721.61 ms/ev initialStepTrackCandidatesPreSplitting
   -0.077693      -0.08%       772.01 ms/ev ->       714.27 ms/ev initialStepTrackCandidates
   -0.075774      -0.11%      1118.65 ms/ev ->      1036.98 ms/ev pixelPairStepTrackCandidates
```
the net effect is about 5.5% of total reco time.

Looking at run2 data workflows (134.902,905,702) there  is about 40MB decrease in RSS (using MemoryCheck).

So, all this is pretty good, except for the conversions efficiency.
Conversions are a small fraction of all tracking: can the selections be relaxed to get the efficiency back?

**Comment by VinInn** [comment]
It is pretty easy (as in MuonStep) to revert the old behaviour for Conversion.
As already noted elsewhere we urgently need an MTV like validation for conversion as I am pretty blind to the real effects of tracking changes to conversions (usually treated as side-effects..)

**Comment by slava77** [comment]
if it's easy, may I suggest to revert to the old behavior for conversions in this PR.

File: RecoTracker/MeasurementDet/plugins/Chi2ChargeMeasurementEstimatorESProducer.cc (hunk 1 above)
Comment (Dr15Jones) [inline RecoTracker/MeasurementDet/plugins/Chi2ChargeMeasurementEstimatorESProducer.cc]: refToPSet is purely a python implementation detail and is unknown by the C++ and therefore would not be useable for validation. I can see how one might want to use it for the default generation of _cfi.py files. My suggestion for this case is to properly describe the required parameters, but do not give them values. Then have a hand written _cfi.py files which is supposed to be used.

**Comment by slava77** [comment]
(still with the previous commit): H->gamma gamma sample (1332, using 600 events) shows some minor increase in conversion efficiency [looking at totals for this plot: 218/443 (49.2 +/- 2.4)% => 226/443 (51.0 +/- 2.4)%]
wf1332_conv_2tk_eff_eta

let's see if the 637019e gives an overall improvement

**Comment by slava77** [comment]
The last change did nothing to the photon gun sample performance, it looks like the loss observed there is either in a different kind of conversion.
In the H->aa sample the last change introduced a reduction of 0.5% in the 2-track conversion efficiency.

So, the last iterations was not that successful.

**Comment by slava77** [comment]
Last time I pointed to IOinnermostHitR plots https://cloud.githubusercontent.com/assets/4676718/12038843/32a5578a-ae10-11e5-8293-c7eb31c2a94f.png . The last commit didn't change these plots at all.

**Comment by slava77** [comment]
Looking at what's used for these tracks, I think we need also
``` chi2CutForConversionTrajectoryBuilder.MaxSagitta = -1 ```

File: RecoTracker/MeasurementDet/plugins/Chi2ChargeMeasurementEstimatorESProducer.cc (hunk 1 above)
Comment (VinInn) [inline RecoTracker/MeasurementDet/plugins/Chi2ChargeMeasurementEstimatorESProducer.cc]: Ok, TRK-POG will address this more globally in a later PR

**Comment by VinInn** [comment]
@cmsbuild, please test

**Comment by VinInn** [comment]
What is the plan to converge in merging this PR?

**Comment by lveldere** [comment]
+1

**Comment by mmusich** [comment]
+1
trivial change in `Alignment/MillePedeAlignmentAlgorithm/python/RefitterAnalytical_cff.py`

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/12846	e0917ef
- code changes look OK; the new propagation logic is applied only to "general" tracks.
- compared to the earlier commits tested and summarized above, the significant differences in physics performance in conversions are now gone
- technical performance: CPU decrease by about 4% in PU35@25ns ttbar  (as https://github.com/cms-sw/cmssw/pull/12846#issuecomment-167847614 but without changes in conversion-related *Candidates)

I suppose, once MTV plots setup for conversions is available, the conversion iterations can be checked again so that we get some speed up there as well.

**Pull request merged**